Create DB properties for internal table cache

matthewvon opened the pull request:
This PR duplicates the existing two properties for BlockCacheCapacity() and BlockCacheUsage() by adding TableCacheCapacity() and TableCacheUsage().

There is currently no way to know how many .sst files are actually open within rocksdb.  The "usage" of the table cache gives that information.  It is important to know how many files are open to estimate the amount of table reader space per open file.  That estimate is then useful for memory management, including the reduction / expansion of the table_cache via SetDBOptions("max_open_files").

mrambacher commented on db/db_properties_test.cc:
```diff
@@ -1713,6 +1716,44 @@ TEST_F(DBPropertiesTest, BlockCacheProperties) {
+  new_options.insert(std::pair<std::string, std::string>(
+      "max_open_files", std::to_string(new_value)));
+  rocksdb::Status stat = db_->SetDBOptions(new_options);
+  ASSERT_TRUE(stat.ok());
```
This should just be ASSERT_OK(db_->SetOptions(...)

matthewvon replied:
Already done.

mrambacher commented on db/db_properties_test.cc:
```diff
@@ -1713,6 +1716,44 @@ TEST_F(DBPropertiesTest, BlockCacheProperties) {
+  ASSERT_EQ(new_value - 10, value);
+
+  //
+  // create a new table to see if usage is "live"
```
I do not see where this is creating a "new table".  Can you elaborate in the comment?

matthewvon replied:
Done.

mrambacher commented on db/db_properties_test.cc:
```diff
@@ -1713,6 +1716,44 @@ TEST_F(DBPropertiesTest, BlockCacheProperties) {
+  options.env = CurrentOptions().env;
+
+  // Block cache properties are not available if block cache is not used.
+  BlockBasedTableOptions table_options;
```
There is already a default options, so you should not need to create the table_options here...

matthewvon replied:
Previous tests make changes.  Thought this was a way to guarantee the traditional .sst tables.  Because I know how to force creation of a new .sst file with the traditional.  Not sure if same rules apply with others.

matthewvon replied:
The redundant table factory creation is now gone.

mrambacher commented on db/db_properties_test.cc:
```diff
@@ -1713,6 +1716,44 @@ TEST_F(DBPropertiesTest, BlockCacheProperties) {
+  rocksdb::Status stat = db_->SetDBOptions(new_options);
+  ASSERT_TRUE(stat.ok());
+  ASSERT_TRUE(db_->GetIntProperty(DB::Properties::kTableCacheCapacity, &value));
+  // rocksdb does a -10 to number passed in
+  ASSERT_EQ(new_value - 10, value);
```
I do not understand how the math works in this test.  You get the value, divide it by 2, subtract 10, and add 1.  By just reading the test this appears to be some sort of odd voodoo.

Can you add some comments somewhere to explain how/why this math works and what you are measuring?

matthewvon replied:
Will add comments.  In short, there are two different properties being tested "capacity" and "usage".

matthewvon commented:
@mrambacher One of the failing tests is upset that I am forcing a compaction without there already being on .sst file.  I will work that in the morning.

matthewvon commented:
@mrambacher The one failing test should succeed.  Changed CompactRange() call to Flush().  Essentially the two calls accomplish the same goal of pushing data from write buffer to a new .sst file.  But somewhere there is code that asserts() if CompactRange() is called with no existing .sst files.  Do not remember which environment.

... and removed the redundant table factory creation in same test.  All should be good now.

mrambacher reviewed (approved):
LGTM
